apollo-router/src/configuration/cors.rs

```diff
@@ -84,90 +154,338 @@ impl Cors {
+
+    fn layer(&self, service: S) -> Self::Service {
+        CorsService {
+            inner: service,
+            config: self.config.clone(),
+        }
+    }
+}
+
+/// Our custom CORS service that handles per-origin configuration
+#[derive(Clone, Debug)]
+pub(crate) struct CorsService<S> {
+    inner: S,
+    config: Cors,
+}
+
+impl<S, ReqBody, ResBody> Service<Request<ReqBody>> for CorsService<S>
+where
+    S: Service<Request<ReqBody>, Response = Response<ResBody>> + Send + 'static,
+    S::Future: Send + 'static,
+    S::Error: Send + 'static,
+    ReqBody: Send + 'static,
+    ResBody: Send + 'static + Default,
+{
+    type Response = S::Response;
+    type Error = S::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Response, Self::Error>> + Send>>;
+
+    fn poll_ready(&mut self, cx: &mut Context<'_>) -> Poll<Result<(), Self::Error>> {
+        self.inner.poll_ready(cx)
+    }
+
+    fn call(&mut self, req: Request<ReqBody>) -> Self::Future {
+        let origin = req.headers().get(ORIGIN).cloned();
+        let is_preflight = req.method() == http::Method::OPTIONS;
+        let config = self.config.clone();
+        let request_headers = req.headers().get("access-control-request-headers").cloned();
+
+        // Intercept OPTIONS requests and return preflight response directly
+        if is_preflight {
+            let mut response = Response::builder()
+                .status(http::StatusCode::OK)
+                .body(ResBody::default())
+                .unwrap();
+            // Find matching origin configuration
+            let origin_config = Self::find_matching_origin_config(&config, &origin);
+            // Add CORS headers for preflight
+            Self::add_cors_headers(
+                &mut response,
+                &config,
+                &origin_config,
+                &origin,
+                true,
+                request_headers,
+            );
+            return Box::pin(async move { Ok(response) });
+        }
+
+        let fut = self.inner.call(req);
+        Box::pin(async move {
+            let mut response = fut.await?;
+            // Find matching origin configuration
+            let origin_config = Self::find_matching_origin_config(&config, &origin);
+            // Add CORS headers for non-preflight
+            Self::add_cors_headers(
+                &mut response,
+                &config,
+                &origin_config,
+                &origin,
+                false,
+                request_headers,
+            );
+            Ok(response)
+        })
+    }
+}
+
+impl<S> CorsService<S> {
+    /// Find the matching origin configuration for a given origin
+    fn find_matching_origin_config<'a>(
+        config: &'a Cors,
+        origin: &'a Option<http::HeaderValue>,
+    ) -> Option<&'a OriginConfig> {
+        let origin_str = origin.as_ref()?.to_str().ok()?;
+
+        // Check for exact origin matches first
+        for origin_config in &config.origins {
```

Comment:
This code will check each origin in order, does it make sense to check all exact matches first before any regex matches? I could see users adding an exact match below a match origin and then not realising that they their rule is not being hit.

Reply:
That could be confusing for users but I did my best to explain it in docs. I'm open to suggestions.

Reply:
SGTM